src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,920 @@
+						errmsg("cannot %s %s after other modifications "
+							   "in the same transaction.",
+							   SplitOperationName[splitOperation],
+							   SplitTargetName[splitOperation])));
+	}
+
+	ShardInterval *shardIntervalToSplit = LoadShardInterval(shardIdToSplit);
+	List *colocatedTableList = ColocatedTableList(shardIntervalToSplit->relationId);
+
+	/* sort the tables to avoid deadlocks */
+	colocatedTableList = SortList(colocatedTableList, CompareOids);
+	Oid colocatedTableId = InvalidOid;
+	foreach_oid(colocatedTableId, colocatedTableList)
+	{
+		/*
+		 * Block concurrent DDL / TRUNCATE commands on the relation. Similarly,
+		 * block concurrent citus_move_shard_placement() / isolate_tenant_to_new_shard()
+		 * on any shard of the same relation.
+		 */
+		LockRelationOid(colocatedTableId, ShareUpdateExclusiveLock);
+	}
+
+	ErrorIfCannotSplitShard(SHARD_SPLIT_API, shardIntervalToSplit);
+	ErrorIfCannotSplitShardExtended(
+		SHARD_SPLIT_API,
+		shardIntervalToSplit,
+		shardSplitPointsList,
+		nodeIdsForPlacementList);
+
+	List *workersForPlacementList = NULL;
+	Datum nodeId;
+	foreach_int(nodeId, nodeIdsForPlacementList)
+	{
+		uint32 nodeIdValue = DatumGetUInt32(nodeId);
+		WorkerNode *workerNode = LookupNodeByNodeId(nodeIdValue);
+
+		/* NodeId in Citus are unsigned and range from [1, 4294967296]. */
+		if (nodeIdValue < 1 || workerNode == NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+							errmsg("Invalid Node Id '%u'.", nodeIdValue)));
+		}
+
+		workersForPlacementList =
+			lappend(workersForPlacementList, (void *) workerNode);
+	}
+
+	if (splitMode == BLOCKING_SPLIT)
+	{
+		EnsureReferenceTablesExistOnAllNodesExtended(TRANSFER_MODE_BLOCK_WRITES);
+		BlockingShardSplit(
+			splitOperation,
+			shardIntervalToSplit,
+			shardSplitPointsList,
+			workersForPlacementList);
+	}
+	else
+	{
+		/* we only support blocking shard split in this code path for now. */
+		ereport(ERROR, (errmsg("Invalid split mode value %d.", splitMode)));
+	}
+}
+
+
+/*
+ * SplitShard API to split a given shard (or shard group) in blocking fashion
+ * based on specified split points to a set of destination nodes.
+ * 'splitOperation'             : Customer operation that triggered split.
+ * 'shardIntervalToSplit'       : Source shard interval to be split.
+ * 'shardSplitPointsList'		: Split Points list for the source 'shardInterval'.
+ * 'workersForPlacementList'	: Placement list corresponding to split children.
+ */
+static void
+BlockingShardSplit(SplitOperation splitOperation,
+				   ShardInterval *shardIntervalToSplit,
+				   List *shardSplitPointsList,
+				   List *workersForPlacementList)
+{
+	List *sourceColocatedShardIntervalList = ColocatedShardIntervalList(
+		shardIntervalToSplit);
+
+	BlockWritesToShardList(sourceColocatedShardIntervalList);
+
+	/* First create shard interval metadata for split children */
+	List *shardGroupSplitIntervalListList = CreateSplitIntervalsForShardGroup(
```

Comment:
I'm wondering whether we can clarify this data structure a little bit, e.g. with some intermediate struct

Reply:
@marcocitus 
I did think about it. We could define an intermediate like below, but did not find it adding too much value.

```
typedef struct ShardSplitIntervals
{
 List *shardSplitIntervalList;
} ShardSplitIntervals;

```
and then return List of ShardSplitIntervals to avoid Lists of List.

Reply:
Keeping it same as of for now.